foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/asm/ASMFactory.java

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright (c) 2023 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+// Contributors:
+//     Oracle - initial API and implementation
+package org.eclipse.persistence.asm;
+
+import org.eclipse.persistence.config.SystemProperties;
+import org.eclipse.persistence.exceptions.ValidationException;
+import org.eclipse.persistence.internal.security.PrivilegedAccessHelper;
+import org.eclipse.persistence.logging.AbstractSessionLog;
+import org.eclipse.persistence.logging.SessionLog;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public class ASMFactory {
+
+    private final static SessionLog LOG = AbstractSessionLog.getLog();
+
+    public final static String ASM_SERVICE_ECLIPSELINK = "eclipselink";
+    public final static String ASM_SERVICE_OW2 = "ow2";
+    private final static String ASM_OW2_CLASS_VISITOR = "org.objectweb.asm.ClassVisitor";
+    private final static String ASM_ECLIPSELINK_CLASS_VISITOR = "org.eclipse.persistence.internal.libraries.asm.ClassVisitor";
+
+    //Should be changed in case of ASM upgrade
+    public final static int ASM_API_SELECTED = Opcodes.valueInt("ASM9");
```

Comment:
So now we're back to having to update Eclipselink every time we need to support a newer level of ASM.  This was one of the strengths of moving the Eclipselink ASM implementation to its own project/bundle.  Is it possible to restore that capability for the path where the Eclipselink ASM bundle is used, and limit this eclipselink binary hard coded constant to the path of using the non-Eclipselink ASM binary?